<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#notes-and-questions" data-toc-modified-id="notes-and-questions-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>notes and questions</a></span></li><li><span><a href="#Setup" data-toc-modified-id="Setup-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Setup---Debug" data-toc-modified-id="Setup---Debug-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Setup - Debug</a></span></li><li><span><a href="#Setup---Imports" data-toc-modified-id="Setup---Imports-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Setup - Imports</a></span></li><li><span><a href="#Setup---working-folder-paths" data-toc-modified-id="Setup---working-folder-paths-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Setup - working folder paths</a></span></li><li><span><a href="#Setup---logging" data-toc-modified-id="Setup---logging-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Setup - logging</a></span></li><li><span><a href="#Setup---virtualenv-jupyter-kernel" data-toc-modified-id="Setup---virtualenv-jupyter-kernel-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Setup - virtualenv jupyter kernel</a></span></li><li><span><a href="#Setup---Initialize-Django" data-toc-modified-id="Setup---Initialize-Django-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Setup - Initialize Django</a></span></li><li><span><a href="#Setup---Initialize-LoggingHelper" data-toc-modified-id="Setup---Initialize-LoggingHelper-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Setup - Initialize LoggingHelper</a></span></li></ul></li><li><span><a href="#Render-network-data" data-toc-modified-id="Render-network-data-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Render network data</a></span><ul class="toc-item"><li><span><a href="#Specify-filter-criteria" data-toc-modified-id="Specify-filter-criteria-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Specify filter criteria</a></span><ul class="toc-item"><li><span><a href="#Old-filter-criteria" data-toc-modified-id="Old-filter-criteria-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Old filter criteria</a></span></li><li><span><a href="#JSON-files-of-filter-criteria-for-just-automated-coding" data-toc-modified-id="JSON-files-of-filter-criteria-for-just-automated-coding-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>JSON files of filter criteria for just automated coding</a></span></li></ul></li><li><span><a href="#Context-network-output" data-toc-modified-id="Context-network-output-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Context network output</a></span></li></ul></li><li><span><a href="#Network-output" data-toc-modified-id="Network-output-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Network output</a></span><ul class="toc-item"><li><span><a href="#Network-output---sourcenet-logic" data-toc-modified-id="Network-output---sourcenet-logic-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Network output - sourcenet logic</a></span><ul class="toc-item"><li><span><a href="#Network-output---sourcenet-logic---NetworkOutput" data-toc-modified-id="Network-output---sourcenet-logic---NetworkOutput-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Network output - sourcenet logic - NetworkOutput</a></span></li><li><span><a href="#Network-output---sourcenet-logic---NetworkDataOutput" data-toc-modified-id="Network-output---sourcenet-logic---NetworkDataOutput-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Network output - sourcenet logic - NetworkDataOutput</a></span></li></ul></li><li><span><a href="#Network-output---context-logic" data-toc-modified-id="Network-output---context-logic-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Network output - context logic</a></span><ul class="toc-item"><li><span><a href="#Network-output---context-logic---NetworkOutput" data-toc-modified-id="Network-output---context-logic---NetworkOutput-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Network output - context logic - NetworkOutput</a></span></li><li><span><a href="#Network-output---context-logic---NetworkDataOutput" data-toc-modified-id="Network-output---context-logic---NetworkDataOutput-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Network output - context logic - NetworkDataOutput</a></span></li></ul></li><li><span><a href="#Context-network-output-test-code" data-toc-modified-id="Context-network-output-test-code-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Context network output test code</a></span></li><li><span><a href="#Building-Unit-Tests" data-toc-modified-id="Building-Unit-Tests-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Building Unit Tests</a></span><ul class="toc-item"><li><span><a href="#unit-test---relation-type-role-value-lists" data-toc-modified-id="unit-test---relation-type-role-value-lists-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>unit test - relation type role value lists</a></span><ul class="toc-item"><li><span><a href="#unit-test---relation-type-role-value-lists---BASIC" data-toc-modified-id="unit-test---relation-type-role-value-lists---BASIC-4.4.1.1"><span class="toc-item-num">4.4.1.1&nbsp;&nbsp;</span>unit test - relation type role value lists - BASIC</a></span></li><li><span><a href="#unit-test---relation-type-role-value-lists---ENTITY_SELECTION" data-toc-modified-id="unit-test---relation-type-role-value-lists---ENTITY_SELECTION-4.4.1.2"><span class="toc-item-num">4.4.1.2&nbsp;&nbsp;</span>unit test - relation type role value lists - ENTITY_SELECTION</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#TODO" data-toc-modified-id="TODO-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>TODO</a></span><ul class="toc-item"><li><span><a href="#TODO---Network-data-creation" data-toc-modified-id="TODO---Network-data-creation-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>TODO - Network data creation</a></span><ul class="toc-item"><li><span><a href="#TODO---Network-data-creation---framework" data-toc-modified-id="TODO---Network-data-creation---framework-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>TODO - Network data creation - framework</a></span></li><li><span><a href="#TODO---Network-data-creation---testing" data-toc-modified-id="TODO---Network-data-creation---testing-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>TODO - Network data creation - testing</a></span></li></ul></li><li><span><a href="#TODO---general" data-toc-modified-id="TODO---general-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>TODO - general</a></span></li></ul></li><li><span><a href="#DONE" data-toc-modified-id="DONE-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>DONE</a></span><ul class="toc-item"><li><span><a href="#DONE---TODO---Network-data-creation" data-toc-modified-id="DONE---TODO---Network-data-creation-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>DONE - TODO - Network data creation</a></span><ul class="toc-item"><li><span><a href="#DONE---TODO---Network-data-creation---framework" data-toc-modified-id="DONE---TODO---Network-data-creation---framework-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>DONE - TODO - Network data creation - framework</a></span></li><li><span><a href="#DONE---TODO---Network-data-creation---testing" data-toc-modified-id="DONE---TODO---Network-data-creation---testing-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>DONE - TODO - Network data creation - testing</a></span></li></ul></li><li><span><a href="#DONE---TODO---general" data-toc-modified-id="DONE---TODO---general-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>DONE - TODO - general</a></span></li></ul></li></ul></div>

# notes and questions

- Back to [Table of Contents](#Table-of-Contents)

Notes:

- probably will need a class that is an Article_Data container - reference to Article_Data, and then all the logic to process the entities and relations contained within.

Questions:

- Do we want an Identifier Type separate from Entity and Relation identifiers?  I think we do, so we can specify the entity type(s) a given identifier should be used on.

    - Created abstract one, and created a concrete entity identifier type.  Will create one for relation identifiers if needed.

# Setup

- Back to [Table of Contents](#Table-of-Contents)

In [1]:
me = "render-context-networks-dev"

## Setup - Debug

- Back to [Table of Contents](#Table-of-Contents)

In [2]:
debug_flag = False

## Setup - Imports

- Back to [Table of Contents](#Table-of-Contents)

In [3]:
import datetime
from django.db.models import Avg, Max, Min, Q
from django.utils.text import slugify
import json
import logging
import six

## Setup - working folder paths

- Back to [Table of Contents](#Table-of-Contents)

In [4]:
%pwd

'/home/jonathanmorgan/work/django/research/work/phd_work/analysis'

In [5]:
# current working folder
current_working_folder = "/home/jonathanmorgan/work/django/research/work/phd_work/analysis"
current_datetime = datetime.datetime.now()
current_date_string = current_datetime.strftime( "%Y-%m-%d-%H-%M-%S" )

## Setup - logging

- Back to [Table of Contents](#Table-of-Contents)

configure logging for this notebook's kernel (If you do not run this cell, you'll get the django application's logging configuration.

In [6]:
logging_file_name = "{}/logs/{}-{}.log.txt".format( current_working_folder, me, current_date_string )
logging.basicConfig(
    level = logging.DEBUG,
    format = '%(asctime)s - %(levelname)s - %(name)s - %(message)s',
    filename = logging_file_name,
    filemode = 'w' # set to 'a' if you want to append, rather than overwrite each time.
)
print( "Logging initialized, to {}".format( logging_file_name ) )

Logging initialized, to /home/jonathanmorgan/work/django/research/work/phd_work/analysis/logs/render-context-networks-dev-2020-02-04-21-16-39.log.txt


## Setup - virtualenv jupyter kernel

- Back to [Table of Contents](#Table-of-Contents)

If you are using a virtualenv, make sure that you:

- have installed your virtualenv as a kernel.
- choose the kernel for your virtualenv as the kernel for your notebook (Kernel --> Change kernel).

Since I use a virtualenv, need to get that activated somehow inside this notebook.  One option is to run `../dev/wsgi.py` in this notebook, to configure the python environment manually as if you had activated the `sourcenet` virtualenv.  To do this, you'd make a code cell that contains:

    %run ../dev/wsgi.py
    
This is sketchy, however, because of the changes it makes to your Python environment within the context of whatever your current kernel is.  I'd worry about collisions with the actual Python 3 kernel.  Better, one can install their virtualenv as a separate kernel.  Steps:

- activate your virtualenv:

        workon research

- in your virtualenv, install the package `ipykernel`.

        pip install ipykernel

- use the ipykernel python program to install the current environment as a kernel:

        python -m ipykernel install --user --name <env_name> --display-name "<display_name>"
        
    `sourcenet` example:
    
        python -m ipykernel install --user --name sourcenet --display-name "research (Python 3)"
        
More details: [http://ipython.readthedocs.io/en/stable/install/kernel_install.html](http://ipython.readthedocs.io/en/stable/install/kernel_install.html)

## Setup - Initialize Django

- Back to [Table of Contents](#Table-of-Contents)

First, initialize my dev django project, so I can run code in this notebook that references my django models and can talk to the database using my project's settings.

In [7]:
# init django
django_init_folder = "/home/jonathanmorgan/work/django/research/work/phd_work"
django_init_path = "django_init.py"
if( ( django_init_folder is not None ) and ( django_init_folder != "" ) ):
    
    # add folder to front of path.
    django_init_path = "{}/{}".format( django_init_folder, django_init_path )
    
#-- END check to see if django_init folder. --#

In [8]:
%run $django_init_path

django initialized at 2020-02-04 21:16:44.765746


In [9]:
# context imports
from context.export.network.filter_spec import FilterSpec
from context.export.network.network_data_request import NetworkDataRequest
from context.export.network.network_output import NetworkOutput
from context.models import Entity
from context.models import Entity_Identifier_Type
from context.models import Entity_Identifier
from context.models import Entity_Relation
from context.models import Entity_Relation_Type
from context.models import Entity_Type
from context.shared.context_base import ContextBase
from context.tests.export.network.test_helper import TestHelper
from context.tests.export.network.test_NetworkDataOutput_class import NetworkDataOutputTest

## Setup - Initialize LoggingHelper

- Back to [Table of Contents](#Table-of-Contents)

Create a LoggingHelper instance to use to log debug and also print at the same time.

Preconditions: Must be run after Django is initialized, since `python_utilities` is in the django path.

In [10]:
# python_utilities
from python_utilities.logging.logging_helper import LoggingHelper

# init
my_logging_helper = LoggingHelper()
my_logging_helper.set_logger_name( me )
log_message = None

# Render network data

- Back to [Table of Contents](#Table-of-Contents)

Now, we need to render out our network data from context, so we can then test it out and make sure we are getting the same answers we got from the old way.

## Specify filter criteria

- Back to [Table of Contents](#Table-of-Contents)

First step is to translate the filter criteria for nodes and ties from the existing admin for the querying context.

### Old filter criteria

- Back to [Table of Contents](#Table-of-Contents)

Configuration of [Network Builder](https://research.local/research/context/text/output/network), from [methods-network_analysis-create_network_data.ipynb](../methods/network_analysis/methods-network_analysis-create_network_data.ipynb):

* Configuration to generate network files for prelim:

    * Config of "Select Articles" - **fields in bold** need to be changed from default values:

        * **`Start date (YYYY-MM-DD):`** 2009-12-01
        * **`End date (YYYY-MM-DD):`** 2009-12-31
        * `Fancy date range:` - Empty.
        * **`Publications:`** "Grand Rapids Press, The"
        * `Coders:` None selected.
        * **`Coder IDs to include, in order of highest to lowest priority:`**

            * for human sample: 13,8,9,10
            * for automated: 2

        * **if automated**: `Article_Data coder_type Filter Type` and `coder_type 'Value In' List (comma-delimited):`
        
            * only for coder "automated" (2), for now.
            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
           
                * **`Article_Data coder_type Filter Type:`** - `Just automated`
                * **`coder_type 'Value In' List (comma-delimited):`** - Enter the coder types you want included.  Examples:
                
                    * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * `Topics`: None selected.
        * **`Article Tag List (comma-delimited):`** - "grp_month"
        * `Unique Identifier List (comma-delimited):` - Empty.
        * **`Allow duplicate articles:`** - "No"
    
    * Configure "Network Settings" - **fields in bold** need to be changed from default values:
        
        * `relations - Include source contact types` - All selected.
        * `relations - Include source capacities:` - None selected.
        * `relations - Exclude source capacities:` - None selected.
        * **`Download as File?`** - "Yes"
        * `Include render details?` - "No"
        * **`Data Format:`** - "Tab-Delimited Matrix"
        * **`Data Output Type:`** - "Network + Attribute Columns"
        * `Network Label:` - Empty.
        * **`Include Headers:`** - "Yes"
    
    * Config of "Select People" - **fields in bold** need to be changed from default values:
        
        * _NOTE: This will be the same for all networks you want to compare (different weeks within a month, compared to the whole month, for instance). For each, get people from articles that are filtered to include all people used by either human or automated, and all the days covered by any of the networks you want to compare.  This means you'll have the same dimensions of network (same set of nodes/people) regardless of the particular network you are generating, allowing the matrices that result to be compared._
        * **`Person Query Type:`** - "Custom, defined below"
        * **`People from (YYYY-MM-DD):`** - 2009-12-01
        * **`People to (YYYY-MM-DD):`** - 2009-12-31
        * `Fancy person date range:` - Empty.
        * **`Person publications:`** - "Grand Rapids Press, The"
        * **`Person coders:`** - "automated", "minnesota1", "minnesota2", "minnesota3", "ground_truth"
        * `Coder IDs to include, in order of highest to lowest priority:` - Empty.
        * **`Article_Data coder_type Filter Type` and `coder_type 'Value In' List (comma-delimited):`**
            
            * _NOTE: not just for automated - since this includes all coders, automated and human, you need to always specify the coder type filter if you need it for automated network._
            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
            
                * **`Article_Data coder_type Filter Type:`** - `Just automated`
                * **`coder_type 'Value In' List (comma-delimited):`** - Enter the coder types you want included.  Examples:
                
                    * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * `Person Topics`: None
        * **`Article Tag List (comma-delimited):`** - "grp_month"
        * `Unique Identifier List (comma-delimited):` - Empty.
        * **`Person allow duplicate articles:`** - "Yes"

### JSON files of filter criteria for just automated coding

- Back to [Table of Contents](#Table-of-Contents)

Below is a JSON file that is just the automated coding for the month from 2009-12-01 through 2009-12-31, articles in the Grand Rapids Press.  Just about all of the complexity of the original screens is possible here, as long as you loaded the entities and ties and all of the needed traits, including some way of adding tags...

    {
        "output_specification": {
            "output_type": "file",
            "output_file_path": "./NetworkDataRequest_test_output.txt",
            "output_format": "TSV_matrix",
            "output_structure": "both_trait_columns",
            "output_include_column_headers": true
        },
        "relation_selection": {
            "relation_type_slug_filter_combine_type": "AND",
            "relation_type_slug_filters": [
                {
                    "comparison_type": "includes",
                    "value_list": [ "mentioned", "qouted", "shared_byline" ]
                }
            ],
            "relation_trait_filter_combine_type": "AND",
            "relation_trait_filters": [
                {
                    "name": "pub_date",
                    "data_type": "date",
                    "comparison_type": "in_range",
                    "value_from": "2009-12-01",
                    "value_to": "2009-12-31"
                },
                {
                    "name": "sourcenet-coder-User-username",
                    "data_type": "string",
                    "comparison_type": "includes",
                    "value_list": [ "automated" ]
                },
                {
                    "name": "coder_type",
                    "data_type": "string",
                    "comparison_type": "includes",
                    "value_list": [ "OpenCalais_REST_API_v2" ]
                }
            ],
            "entity_type_slug_filter_combine_type": "AND",
            "entity_type_slug_filters": [
                {
                    "comparison_type": "includes",
                    "value_list": [ "person" ],
                    "relation_roles_list": [ "FROM" ]
                },
                {
                    "comparison_type": "includes",
                    "value_list": [ "person" ],
                    "relation_roles_list": [ "TO" ]
                },
                {
                    "comparison_type": "includes",
                    "value_list": [ "article" ],
                    "relation_roles_list": [ "THROUGH" ]
                }            
            ],
            "entity_trait_filter_combine_type": "AND",
            "entity_trait_filters": [
                {
                    "name": "sourcenet-Newspaper-ID",
                    "data_type": "int",
                    "comparison_type": "includes",
                    "value_list": [ 1 ],
                    "relation_roles_list": [ "THROUGH" ]
                }
            ]
        }
    }

## Context network output

- Back to [Table of Contents](#Table-of-Contents)

Steps:

- make sure `settings.py` is referring to the full database (`research`, not the testing database `research_temp`)
- run all the initialization code above in section 2.
- load a request into a NetworkDataRequest instance (example below uses the unit test TestHelper class to load the basic configuration, which is the same as the above, for 12/1/2009 through 12/31/2009, automated coder, Grand Rapids Press only).

    - using TestHelper: `data_request = TestHelper.load_basic()`
    - from file path:
    
            data_request = NetworkDataRequest()
            path_to_file = "./request.json"
            data_request.load_network_data_request_json_file( path_to_file )

- update details of the request post-load.  Examples:

    - call "`data_request.set_output_file_path( path )`" to change where the file will be output.
    - call "`data_request.set_output_format( format )`" to change the output format.  Values:
    
        - "simple_matrix" ( `NetworkDataRequest.PROP_VALUE_OUTPUT_FORMAT_SIMPLE_MATRIX` ) - format for UCINet where contents of tie file and attribute files are all concatenated into one text file.
        - "CSV_matrix" ( `NetworkDataRequest.PROP_VALUE_OUTPUT_FORMAT_CSV_MATRIX` )
        - "TSV_matrix" ( `NetworkDataRequest.PROP_VALUE_OUTPUT_FORMAT_TSV_MATRIX` )
    
    - call "`data_request.set_output_type( type )`" to change the output type. 

- make an instance of NetworkOutput: `network_output = NetworkOutput()`.
- initialize it by calling "`network_output.set_network_data_request( data_request )`", passing in the request you loaded.
- call "`network_output.render_network_data()`".  The method returns the contents of the file it created, and outputs to file as directed inside the request.

In [14]:
# load basic NetworkDataRequest
#data_request = TestHelper.load_basic()

# or, load from file path
data_request = NetworkDataRequest()
path_to_file = "./grp_month_from_context.json"
data_request.load_network_data_request_json_file( path_to_file )

In [12]:
# configure output
#overridden_output_path = "{}/automated_grp_month.tsv".format( current_working_folder )
#data_request.set_output_file_path( overridden_output_path )

'/home/jonathanmorgan/work/django/research/work/phd_work/analysis/automated_grp_month.tsv'

In [15]:
# make and initialize instance of NetworkOutput
network_output = NetworkOutput()
network_output.set_network_data_request( data_request )

In [16]:
# call render, see what happens.
network_data = network_output.render_network_data()

In context.export.network.network_output.NetworkOutput.render_network_data(): Calling NetworkDataRequest.process_entities() @ 2020-02-04 21:37:14.739044
In get_selection_filters(): use_entity_selection_IN is True
In get_selection_filters(): "entity_selection" filtering was requested, but not specified in the request.  Defaulting to "relation_selection".
In context.export.network.network_output.NetworkOutput.render_network_data(): After calling NetworkDataRequest.process_entities(), calling NetworkDataRequest.filter_relation_query_set() @ 2020-02-04 21:37:45.084237
In get_selection_filters(): use_entity_selection_IN is False
In context.export.network.network_output.NetworkOutput.render_network_data(): After calling NetworkDataRequest.filter_relation_query_set(), initializing NetworkDataOutput instance @ 2020-02-04 21:37:45.090883
In context.export.network.network_output.NetworkOutput.render_network_data(): After initializing NetworkDataOutput instance, calling NetworkDataOutput.render()

In context.export.network.network_output.NetworkOutput.render_network_data(): After calling NetworkDataOutput.render(), doing output @ 2020-02-04 21:39:17.897726
In context.export.network.network_output.NetworkOutput.render_network_data(): Finished @ 2020-02-04 21:39:17.900605


In [14]:
ndo_instance = network_output.m_NDO_instance
test1 = ndo_instance.get_entity_relation_type_summary_dict()
print( test1 )

test2 = ndo_instance.m_relation_type_slug_list
print( "\n\nEntity_Relation_Type slug list: {}".format( test2 ) )

test3 = ndo_instance.m_relation_type_slug_to_instance_map
print( "\n\nEntity_Relation_Type slug to instance map: {}".format( test3 ) )

{35: {'quoted': {'FROM': 1, 'TO': 0, 'THROUGH': 0}, 'mentioned': {'FROM': 1, 'TO': 0, 'THROUGH': 0}}, 36: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 18: {'quoted': {'FROM': 6, 'TO': 0, 'THROUGH': 0}, 'mentioned': {'FROM': 7, 'TO': 0, 'THROUGH': 0}}, 24: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 23: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 22: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 21: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 20: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 19: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 25: {'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 8: {'shared_byline': {'FROM': 1, 'T

# Network output

- Back to [Table of Contents](#Table-of-Contents)

Convert code from using Article_Data to derive ties to just loop over Entity_Relation instances.

## Network output - sourcenet logic

- Back to [Table of Contents](#Table-of-Contents)


### Network output - sourcenet logic - NetworkOutput

- Back to [Table of Contents](#Table-of-Contents)

Pseudocode:

- NetworkOutput.render_network_data( query_set_IN )

    - accepts Article_Data QuerySet.
    - create person dictionary - self.create_person_dict()
    
        - retrieve Article_Data QuerySet to traverse to pull in all authors and subjects - self.create_person_query_set()
        
            - if ALL, return all Article_Data
            - if type of articles, retrieve just those specified as being used for ties - self.create_person_query_set()
            - if type of custom, retrieve Article_Data to match custom person lookup parameters - self.create_query_set( NetworkOutput.PARAM_PERSON_PREFIX )
            
        - for each Article_Data:
        
            - retrieve author and source QuerySets.
            - for each, call self.add_people_to_dict() on the QuerySet.
            
                - for each person, retrieve person instance and ID, add to dictionary, storing instance as value if store_person_IN == True.
                
        - return dictionary into "person_dictionary".
    
    - get NetworkDataOutput (NDO) instance - self.get_NDO_instance()
    - initialize NDO
    
        - NDO.set_query_set( query_set_IN )
        - NDO.set_person_dictionary( person_dictionary )
        - initialize from parameters: NDO.initialize_from_params()
        
    - render - NDO.render()
    

### Network output - sourcenet logic - NetworkDataOutput

- Back to [Table of Contents](#Table-of-Contents)

Pseudocode:

- NetworkDataOutput.render()

    - get query_set and person_dictionary.
    - create ties

        - loop over query_set:
        
            - check if authors.  If none, move on.
            - if authors:
            
                - retrieve author and sources QuerySets from Article_Data
                - process author relations (shared byline and qouted): `self.process_author_relations( author_qs, source_qs )`
                
                    - makes list of author IDs in Article_Data by building dictionary that maps author person ID to author Person instance, then getting `.keys()`.
                    - loops over authors:
                    
                        - for each, pop from dictionary, make relation to all authors remaining in the dictionary (bi-directional, so just need to make relations to people past the current in the list): `self.add_reciprocal_relation()`
                        
                            - adds bidirectional ties between authors in nested connection map (calls `self.add_directed_relation( person_1_id_IN, person_2_id_IN )`, then `self.add_directed_relation( person_2_id_IN, person_1_id_IN )`).
                            
                                - `self.add_directed_relation()`: updates `self.relation_map`, map of FROM IDs to dictionary of TO IDs, where each TO maps to a count of the ties between the two.  **_Solely ID-based, so methods to add relations can stay as-is!  Yay!_**
                        
                        - set person's type to "author": `self.update_person_type( current_person_id, NetworkDataOutput.PERSON_TYPE_AUTHOR )`
                        
                            - accepts person ID and type.
                            - looks for person ID in `self.person_type_dict`.
                            
                                - If not present, adds current type.
                                - If present but not same as what is passed in, sets to BOTH.
                        
                        - add relations to sources: `self.process_source_relations( current_person_id, source_qs_IN )`
                        
                            - accepts author ID and source QuerySet.
                            - if author ID and source QuerySet has something in it, proceeds.
                            - checks if source is connected: `self.is_source_connected( current_source )`
                            
                                - calls `Article_Subject.is_connected( self.inclusion_params )`
                            
                            - if connected, retrieves person ID for source, if source has ID adds reciprocal relation between author and source: `self.add_reciprocal_relation()`.
                            
                - update the types of the sources (from source_qs): `self.update_source_person_types( source_qs )`
                
                    - for each source, calls `self.update_person_type( current_person_id, NetworkDataOutput.PERSON_TYPE_SOURCE )`
                
        - build master person list: `self.generate_master_person_list()`
        
            - combines the person dictionary with dictionary that maps person to source type based on Article_Data processed for ties.
            
        - actually render the network data: `self.render_network_data()`
        
            - based on output parameters, implemented in NDO child classes.

## Network output - context logic

- Back to [Table of Contents](#Table-of-Contents)


### Network output - context logic - NetworkOutput

- Back to [Table of Contents](#Table-of-Contents)

Pseudocode:

- NetworkOutput.render_network_data( relation_qs_IN, network_data_request_IN )

    - accepts optional Entity_Relation QuerySet and optional request.
    - create entity dictionary - self.create_entity_dict()
    
        - retrieve Entity_Relation QuerySet to traverse to pull in all relations - network_data_request.filter_relation_query_set( use_entity_selection_IN = True )
        - call self.add_entities_to_dict() to add entities from the relation QuerySet.
        
            - loops over relations, retrieves FROM, TO, and optionally THROUGH entities, calls self.add_entity_to_dict() on each entity to be added.
            
                - for each entity, retrieve Entity instance and ID, add to dictionary, storing instance as value if store_entity_IN == True.
                
        - return dictionary into "entity_dictionary".
    
    - get NetworkDataOutput (NDO) instance - self.get_NDO_instance()
    - initialize NDO
    
        - NDO.set_query_set( query_set_IN )
        - NDO.set_entity_dictionary( entity_dictionary )
        - initialize from request: NDO.initialize_from_request( network_data_request )
        
    - render - NDO.render()
    

### Network output - context logic - NetworkDataOutput

- Back to [Table of Contents](#Table-of-Contents)

Pseudocode:

- NetworkDataOutput.render()

    - get Entity_Relation query_set and entity_dictionary.
    - create ties

        - loop over query_set:
        
            - // retrieve FROM and TO entities.
            - // pull out their IDs.
            - // add relation - check if directed:
            
                - directed:
                
                    - `self.add_directed_relation( from_id, to_id )`
                    
                - not directed:
                
                    - `self.add_reciprocal_relation( from_id, to_id )`
            
            - // update roles of entities...
            
                - one thought is to make a map of relation types for each person that maps to a dictionary of FROM, TO, and THROUGH, each mapped to counts of the times the person was in each role.
                - replaces:

                    - calls to: `self.update_person_type( current_person_id, NetworkDataOutput.PERSON_TYPE_SOURCE )` with `update_entity_relations_details( self, entity_id_IN, relation_type_slug_IN, relation_role_IN ):`
                    - references to `self.person_type_dict` with something more nuanced (no longer can depend on a single type).
                
        - // build master entity list: `self.generate_master_entity_list()`
        
            - combines the person dictionary with dictionary that maps person to source type based on Article_Data processed for ties.
            - should be able to use it as is with some variable name changes.
            - // also, change name to "m_master_entity_list", and make a getter(), and then always retrieve using get().
            
        - actually render the network data: `self.render_network_data()`
        
            - based on output parameters, implemented in NDO child classes.

## Context network output test code

- Back to [Table of Contents](#Table-of-Contents)

Try loading the basic file and using it to render network data...

In [15]:
# load basic NetworkDataRequest
data_request_basic = TestHelper.load_basic()

In [16]:
# make and initialize instance of NetworkOutput
network_output = NetworkOutput()
network_output.set_network_data_request( data_request_basic )

In [17]:
# call render, see what happens.
network_data = network_output.render_network_data()

In get_selection_filters(): use_entity_selection_IN is True
In get_selection_filters(): "entity_selection" filtering was requested, but not specified in the request.  Defaulting to "relation_selection".
In get_selection_filters(): use_entity_selection_IN is False
In generate_master_entity_list: len( entity_dict ) = 72; entity_dict: {35: None, 36: None, 18: None, 24: None, 23: None, 22: None, 21: None, 20: None, 19: None, 25: None, 8: None, 9: None, 16: None, 13: None, 12: None, 11: None, 15: None, 14: None, 10: None, 80: None, 83: None, 82: None, 81: None, 69: None, 72: None, 71: None, 70: None, 46: None, 44: None, 40: None, 47: None, 45: None, 43: None, 42: None, 41: None, 39: None, 38: None, 32: None, 28: None, 33: None, 31: None, 30: None, 29: None, 27: None, 85: None, 89: None, 87: None, 90: None, 88: None, 86: None, 74: None, 78: None, 77: None, 76: None, 75: None, 63: None, 67: None, 66: None, 65: None, 64: None, 49: None, 53: None, 52: None, 51: None, 50: None, 55: None, 61: Non

In [14]:
ndo_instance = network_output.m_NDO_instance
test1 = ndo_instance.get_entity_relation_type_summary_dict()
print( test1 )

test2 = ndo_instance.m_relation_type_slug_list
print( "\n\nEntity_Relation_Type slug list: {}".format( test2 ) )

test3 = ndo_instance.m_relation_type_slug_to_instance_map
print( "\n\nEntity_Relation_Type slug to instance map: {}".format( test3 ) )

{35: {'quoted': {'FROM': 1, 'TO': 0, 'THROUGH': 0}, 'mentioned': {'FROM': 1, 'TO': 0, 'THROUGH': 0}}, 36: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 18: {'quoted': {'FROM': 6, 'TO': 0, 'THROUGH': 0}, 'mentioned': {'FROM': 7, 'TO': 0, 'THROUGH': 0}}, 24: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 23: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 22: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 21: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 20: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 19: {'quoted': {'FROM': 0, 'TO': 1, 'THROUGH': 0}, 'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 25: {'mentioned': {'FROM': 0, 'TO': 1, 'THROUGH': 0}}, 8: {'shared_byline': {'FROM': 1, 'T

## Building Unit Tests

- Back to [Table of Contents](#Table-of-Contents)

Code to test out and find answers for unit tests.

In [17]:
# randomly pick 5 or so relations that have THROUGH.
relation_qs = Entity_Relation.objects.all()
relation_qs = relation_qs.exclude( relation_through__isnull = True )
relation_qs = relation_qs.order_by('?')[:5]
relation_count = len( relation_qs )
print( "relation count: {}".format( relation_count ) )

relation count: 5


In [21]:
# randomly pick 5 or so relations that have THROUGH.
relation_qs = Entity_Relation.objects.all()
relation_qs = relation_qs.exclude( pk__in = [ 1101, 1245, 2264, 2687, 2973 ] )
relation_qs = relation_qs.filter( relation_from__in = [ 80, 128, 147, 151, 246, 248, 271, 279, 288, 291 ] )
relation_qs = relation_qs.order_by('?')[:5]
relation_count = len( relation_qs )
print( "relation count: {}".format( relation_count ) )

# loop
for relation in relation_qs:
    
    print( "Relation ID: {}".format( relation.id ) )
    
#-- END loop over relations --#

relation count: 5
Relation ID: 2638
Relation ID: 3086
Relation ID: 2271
Relation ID: 2595
Relation ID: 2288


In [23]:
relation_id_list = [ 1101, 1245, 2264, 2271, 2288, 2595, 2638, 2687, 2973, 3086 ]
#relation_id_list = TestHelper.TEST_RELATION_IDS_WITH_THROUGH
relation_qs = Entity_Relation.objects.all()
relation_qs = relation_qs.filter( pk__in = relation_id_list )

print( "relation count: {}".format( relation_count ) )

# loop
for relation in relation_qs:
    
    print( "Relation ID: {}".format( relation.id ) )
    
#-- END loop over relations --#

relation count: 5
Relation ID: 1101
Relation ID: 1245
Relation ID: 2264
Relation ID: 2271
Relation ID: 2288
Relation ID: 2595
Relation ID: 2638
Relation ID: 2687
Relation ID: 2973
Relation ID: 3086


In [25]:
# declare variables
relation_id_list = None
relation = None
relation_id = None
relation_from = None
relation_from_id = None
relation_to = None
relation_to_id = None
relation_through = None
relation_through_id = None
from_to_entity_id_list = None
all_entity_id_list = None
relation_counter = None

# init
relation_id_list = []
from_to_entity_id_list = []
all_entity_id_list = []

# loop
relation_counter = 0
for relation in relation_qs:
    
    # increment counter
    relation_counter += 1
    
    # get relation ID
    relation_id = relation.id
    
    # add it to list?
    if ( relation_id not in relation_id_list ):
    
        # not already in list.  Add it.
        relation_id_list.append( relation_id )
        relation_id_list.sort()
    
    #-- END check to see if ID already in list. --#
    
    # get entities
    relation_from = relation.relation_from
    relation_to = relation.relation_to
    relation_through = relation.relation_through
    
    # get IDs, check for each in entity ID lists
    relation_from_id = relation_from.id

    if ( relation_from_id not in from_to_entity_id_list ):
        
        # append it and sort.
        from_to_entity_id_list.append( relation_from_id )
        from_to_entity_id_list.sort()
        
    #-- END check to see if FROM in FROM-TO list --#
    
    if ( relation_from_id not in all_entity_id_list ):
        
        # append it and sort.
        all_entity_id_list.append( relation_from_id )
        all_entity_id_list.sort()
        
    #-- END check to see if FROM in FROM-TO list --#
    
    relation_to_id = relation_to.id

    if ( relation_to_id not in from_to_entity_id_list ):
        
        # append it and sort.
        from_to_entity_id_list.append( relation_to_id )
        from_to_entity_id_list.sort()
        
    #-- END check to see if FROM in FROM-TO list --#
    
    if ( relation_to_id not in all_entity_id_list ):
        
        # append it and sort.
        all_entity_id_list.append( relation_to_id )
        all_entity_id_list.sort()
        
    #-- END check to see if FROM in FROM-TO list --#
    
    relation_through_id = relation_through.id
    
    if ( relation_through_id not in all_entity_id_list ):
        
        # append it and sort.
        all_entity_id_list.append( relation_through_id )
        all_entity_id_list.sort()
        
    #-- END check to see if FROM in FROM-TO list --#
    
    print( "\n- after relation #{} - {}:".format( relation_counter, relation ) )
    print( "----> Relation ID List ( {} ): {}".format( len( relation_id_list ), relation_id_list ) )
    print( "----> from_to_entity_id_list ( {} ): {}".format( len( from_to_entity_id_list ), from_to_entity_id_list ) )
    print( "----> all_entity_id_list ( {} ): {}".format( len( all_entity_id_list ), all_entity_id_list ) )

#-- END loop over Relations --#

print( "Relation ID List ( {} ): {}".format( len( relation_id_list ), relation_id_list ) )
print( "from_to_entity_id_list ( {} ): {}".format( len( from_to_entity_id_list ), from_to_entity_id_list ) )
print( "all_entity_id_list ( {} ): {}".format( len( all_entity_id_list ), all_entity_id_list ) )


- after relation #1 - 1101 - FROM: 80 - context_text-Person-591 - TO: 128 - context_text-Person-1044 - THROUGH: 125 - context_text-Article-28846 - DIRECTED?: False - ( type: 2 - mentioned ):
----> Relation ID List ( 1 ): [1101]
----> from_to_entity_id_list ( 2 ): [80, 128]
----> all_entity_id_list ( 3 ): [80, 125, 128]

- after relation #2 - 1245 - FROM: 147 - context_text-Person-56 - TO: 151 - context_text-Person-170 - THROUGH: 146 - context_text-Article-91000 - DIRECTED?: False - ( type: 2 - mentioned ):
----> Relation ID List ( 2 ): [1101, 1245]
----> from_to_entity_id_list ( 4 ): [80, 128, 147, 151]
----> all_entity_id_list ( 6 ): [80, 125, 128, 146, 147, 151]

- after relation #3 - 2264 - FROM: 246 - context_text-Person-1016 - TO: 248 - context_text-Person-905 - THROUGH: 244 - context_text-Article-94125 - DIRECTED?: False - ( type: 9 - same_article_subjects ):
----> Relation ID List ( 3 ): [1101, 1245, 2264]
----> from_to_entity_id_list ( 6 ): [80, 128, 147, 151, 246, 248]
----> 

### unit test - relation type role value lists

- Back to [Table of Contents](#Table-of-Contents)

#### unit test - relation type role value lists - BASIC

- Back to [Table of Contents](#Table-of-Contents)

In [12]:
# make test instance.
test_instance = NetworkDataOutputTest()

# set up basic
basic_instance = test_instance.set_up_basic_test_instance()

# render
basic_instance.render()

# get relation QuerySet
relation_qs = basic_instance.get_query_set()

# get master ID list
entity_id_list = basic_instance.get_master_entity_list()

In get_selection_filters(): use_entity_selection_IN is True
In get_selection_filters(): "entity_selection" filtering was requested, but not specified in the request.  Defaulting to "relation_selection".
In get_selection_filters(): use_entity_selection_IN is False
In generate_master_entity_list: len( entity_dict ) = 72; entity_dict: {35: None, 36: None, 18: None, 24: None, 23: None, 22: None, 21: None, 20: None, 19: None, 25: None, 8: None, 9: None, 16: None, 13: None, 12: None, 11: None, 15: None, 14: None, 10: None, 80: None, 83: None, 82: None, 81: None, 69: None, 72: None, 71: None, 70: None, 46: None, 44: None, 40: None, 47: None, 45: None, 43: None, 42: None, 41: None, 39: None, 38: None, 32: None, 28: None, 33: None, 31: None, 30: None, 29: None, 27: None, 85: None, 89: None, 87: None, 90: None, 88: None, 86: None, 74: None, 78: None, 77: None, 76: None, 75: None, 63: None, 67: None, 66: None, 65: None, 64: None, 49: None, 53: None, 52: None, 51: None, 50: None, 55: None, 61: Non

In [16]:
test_entity = Entity.objects.get( id = 18 )
print( test_entity )

test_qs = relation_qs.filter( relation_through = test_entity )
print( "match count: {}".format( test_qs.count() ) )

test_qs = relation_qs.filter( relation_through_id = 18 )
print( "match count: {}".format( test_qs.count() ) )

18 - context_text-Person-30
match count: 0
match count: 0


In [21]:
mentioned_type = Entity_Relation_Type.objects.get( slug = "mentioned" )
quoted_type = Entity_Relation_Type.objects.get( slug = "quoted" )
shared_byline_type = Entity_Relation_Type.objects.get( slug = "shared_byline" )
type_list = [ mentioned_type, quoted_type, shared_byline_type ]
role_list = [ ContextBase.RELATION_ROLES_FROM, ContextBase.RELATION_ROLES_TO, ContextBase.RELATION_ROLES_THROUGH ]

# declare variables
current_type_slug = None
type_to_roles_map = {}
role_to_list_map = {}

# loop over types
for current_type in type_list:
    
    # add to type_to_roles_map
    current_type_slug = current_type.slug
    role_to_list_map = {}
    type_to_roles_map[ current_type_slug ] = role_to_list_map
    
    # loop over roles
    for current_role in role_list:
        
        # init list
        type_role_list = []
        
        # loop over IDs.
        for entity_id in entity_id_list:
            
            # do lookup in relations.  relation type...
            test_qs = relation_qs.filter( relation_type = current_type )

            # and entity in requested role...
            if ( current_role == ContextBase.RELATION_ROLES_FROM ):
                test_qs = test_qs.filter( relation_from_id = entity_id )
            elif ( current_role == ContextBase.RELATION_ROLES_TO ):
                test_qs = test_qs.filter( relation_to_id = entity_id )
            elif ( current_role == ContextBase.RELATION_ROLES_THROUGH ):
                test_qs = test_qs.filter( relation_through_id = entity_id )
            #-- END check to see which role we are filtering on. --#
            
            # get count of matches for this entity, requested type and role
            entity_count = test_qs.count()
            
            # add count to list.
            type_role_list.append( entity_count )
            
        #-- END loop over entities. --#
        
        # output the list.
        print( "----> type: {}; role: {}; list: {}".format( current_type, current_role, type_role_list ) )
        
        # add it to role map for current type
        role_to_list_map[ current_role ] = type_role_list
        
    #-- END loop over roles --#
    
#-- END loop over relation types. --#

----> type: 2 - mentioned - mentioned; role: FROM; list: [7, 24, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 3, 0, 0, 0, 4, 0, 0, 0, 0, 3, 0, 0, 0, 5, 0, 0, 0, 0, 0]
----> type: 2 - mentioned - mentioned; role: TO; list: [0, 0, 2, 2, 2, 2, 2, 2, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1]
----> type: 2 - mentioned - mentioned; role: THROUGH; list: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
----> type: 1 - quoted - quoted; role: FROM; list: [4, 9, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 3

#### unit test - relation type role value lists - ENTITY_SELECTION

- Back to [Table of Contents](#Table-of-Contents)

In [22]:
# make test instance.
test_instance = NetworkDataOutputTest()

# set up basic
ndo_instance = test_instance.set_up_entity_selection_test_instance()

# render
ndo_instance.render()

# get relation QuerySet
relation_qs = ndo_instance.get_query_set()

# get master ID list
entity_id_list = ndo_instance.get_master_entity_list()

In get_selection_filters(): use_entity_selection_IN is True
In get_selection_filters(): use_entity_selection_IN is False
In generate_master_entity_list: len( entity_dict ) = 153; entity_dict: {135: None, 140: None, 139: None, 138: None, 136: None, 137: None, 161: None, 209: None, 218: None, 217: None, 216: None, 215: None, 214: None, 213: None, 212: None, 211: None, 72: None, 157: None, 210: None, 162: None, 163: None, 188: None, 190: None, 189: None, 171: None, 176: None, 175: None, 174: None, 173: None, 172: None, 35: None, 36: None, 220: None, 221: None, 222: None, 142: None, 144: None, 145: None, 143: None, 18: None, 24: None, 23: None, 22: None, 21: None, 20: None, 19: None, 25: None, 178: None, 181: None, 180: None, 185: None, 184: None, 183: None, 182: None, 179: None, 8: None, 9: None, 16: None, 13: None, 12: None, 11: None, 15: None, 14: None, 10: None, 80: None, 83: None, 82: None, 81: None, 192: None, 201: None, 200: None, 195: None, 194: None, 199: None, 198: None, 197: Non

In [23]:
test_entity = Entity.objects.get( id = 18 )
print( test_entity )

test_qs = relation_qs.filter( relation_through = test_entity )
print( "match count: {}".format( test_qs.count() ) )

test_qs = relation_qs.filter( relation_through_id = 18 )
print( "match count: {}".format( test_qs.count() ) )

18 - context_text-Person-30
match count: 0
match count: 0


In [24]:
mentioned_type = Entity_Relation_Type.objects.get( slug = "mentioned" )
quoted_type = Entity_Relation_Type.objects.get( slug = "quoted" )
shared_byline_type = Entity_Relation_Type.objects.get( slug = "shared_byline" )
type_list = [ mentioned_type, quoted_type, shared_byline_type ]
role_list = [ ContextBase.RELATION_ROLES_FROM, ContextBase.RELATION_ROLES_TO, ContextBase.RELATION_ROLES_THROUGH ]

# declare variables
current_type_slug = None
entity_selection_type_to_roles_map = {}
role_to_list_map = {}

# loop over types
for current_type in type_list:
    
    # add to type_to_roles_map
    current_type_slug = current_type.slug
    role_to_list_map = {}
    entity_selection_type_to_roles_map[ current_type_slug ] = role_to_list_map
    
    # loop over roles
    for current_role in role_list:
        
        # init list
        type_role_list = []
        
        # loop over IDs.
        for entity_id in entity_id_list:
            
            # do lookup in relations.  relation type...
            test_qs = relation_qs.filter( relation_type = current_type )

            # and entity in requested role...
            if ( current_role == ContextBase.RELATION_ROLES_FROM ):
                test_qs = test_qs.filter( relation_from_id = entity_id )
            elif ( current_role == ContextBase.RELATION_ROLES_TO ):
                test_qs = test_qs.filter( relation_to_id = entity_id )
            elif ( current_role == ContextBase.RELATION_ROLES_THROUGH ):
                test_qs = test_qs.filter( relation_through_id = entity_id )
            #-- END check to see which role we are filtering on. --#
            
            # get count of matches for this entity, requested type and role
            entity_count = test_qs.count()
            
            # add count to list.
            type_role_list.append( entity_count )
            
        #-- END loop over entities. --#
        
        # output the list.
        print( "----> type: {}; role: {}; list: {}".format( current_type, current_role, type_role_list ) )
        
        # add it to role map for current type
        role_to_list_map[ current_role ] = type_role_list
        
    #-- END loop over roles --#
    
#-- END loop over relation types. --#

----> type: 2 - mentioned - mentioned; role: FROM; list: [7, 24, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 3, 0, 0, 0, 4, 0, 0, 0, 0, 3, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
----> type: 2 - mentioned - mentioned; role: TO; list: [0, 0, 2, 2, 2, 2, 2, 2, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

# TODO

- Back to [Table of Contents](#Table-of-Contents)

## TODO - Network data creation

- Back to [Table of Contents](#Table-of-Contents)

### TODO - Network data creation - framework

- Back to [Table of Contents](#Table-of-Contents)

Framework todos:

- add in a little more detailed updates on rendering network data (every X relations, output timestamp).
- build basic framework where each output type accepts these two QuerySets, renders and returns desired output format.

    - have child NDO classes implement "`initialize_from_request`" if needed, call parent, then init format-specific stuff.
    - update to use StatusContainer instead of status strings.

- add type checking to setters of dictionaries and lists, to make sure either None or correct type passed in.

### TODO - Network data creation - testing

- Back to [Table of Contents](#Table-of-Contents)

Testing todos:

- // first, get it running, then unit tests.
- // make plan for unit testing for all export instances, fill in outline below.
- find a way to put method `validate_string_against_file_contents()` into python_utilities - new unittest_helper class, and call the stock unittest asserts instead of those on the instance?

    - make a parent class that extends the base unittest class? (`unittest.TestCase`)
    - // make a parent class that extends Django's unit test class (`django.test.TestCase`).

- unit testing:

    - NetworkDataRequest
    
        - // `do_output_entity_traits_or_ids`
        - // class method `create_entity_id_header_label`
        - // class method `create_entity_trait_header_label`
        - // `create_entity_ids_and_traits_header_list`
        - // `get/set_entity_ids_and_traits_header_list`
        - // `get/set_entity_id_to_instance_map`
        - // `get/set_entity_id_to_traits_map`
        - // `process_entities` - add to test to include entity traits and IDs

            - // make sure structure and count are right
            - add test for entity ID list?
            - // spot-check a few individuals for values
            
                - 8 - 161; None; Nardy; Baeza; Bickel
                - 27 - 877; http://d.opencalais.com/pershash-1/15e6a7c0-40fa-3235-bbec-f46f2d362782; Kaidon; None; None
                - 57 - 933; http://d.opencalais.com/pershash-1/5de8e1c8-9f6b-32de-b77b-29dab0a9c3c2; Douglas; Vander; Hart
                - 66 - 274; http://d.opencalais.com/pershash-1/18bc2abb-9540-300a-b6e1-2f75366848bd; Stacie; None; Behler
                - 85 - 84; None; John; None; Tunison
            
            - build tests for the following:

                - // `load_entities_ids_and_traits`
                - // `load_entity_identifiers`
                - // `load_entity_traits`

        - **NOTE: Add `process_entities` as precondition for the following:**    
        - // `create_ids_and_traits_values_for_entity`

            - // `get_ids_and_traits_for_entity`

        - // `get/set_entity_id_list`
        - // `generate_entity_id_list`
        - `create_entity_ids_and_traits_value_dict( entity_id_list_IN )`

            - loop over labels, calling the validate method for `create_entity_ids_and_traits_value_list` for each.
            - `create_entity_ids_and_traits_value_list( self, header_label_IN, entity_id_list_IN = None ):` - make a "validate" method that accepts...? and:
            
                - check default list
                - check a smaller list
                - check a smaller list that includes entities who were not loaded
            
        - `load_ids_and_traits_for_entities( self, entity_id_list_IN, dictionary_IN )`
        - `process_entities_from_id_list`

    - NetworkDataOutput
    
        - dependencies for child classes:

            - `create_header_list()`
            - `create_label_list()`
            - `create_relation_type_roles_for_entity()`
            - `create_relation_type_roles_header_list()`
            - `do_output_attribute_columns()`
            - `do_output_attribute_rows()`
            - `do_output_network()`
            - `get_entity_label()`
            - `get_relation_roles_for_entity()`
            - `get_relations_for_entity()`

        - notes

            - `register_relation_type()`, and the places that call it: `render()`, optionally also `update_entity_relation_details()`

    - NDO children

        - NDO_SimpleMatrix

            - `render_network_data()` - at a high level, render the basic, compare to a pre-rendered file.

                - `create_label_string()`

            - `create_network_string()` - the worker method, effectively, testing `render_network_data()` tests this.

                - `create_entity_row_string()` - per row method.

            - `create_entity_relation_types_attribute_string()`

        - NDO_CSVMatrix

            - `append_entity_ids_and_traits_rows`
            - `render_network_data()` - at a high level, render the basic, compare to a pre-rendered file.

                - `create_csv_string()`

                    - `init_csv_output()`
                    - `create_csv_document()`

                        - `create_header_list()`
                        - `append_row_to_csv()`
                        - `append_entity_row()`

                            - `create_relation_type_roles_for_entity()`
                            - (duplicate) `append_row_to_csv()`

                        - `append_entity_id_row()`

                            - (duplicate) `append_row_to_csv()`

                        - `append_entity_relation_type_rows()`

                            - (duplicate) `append_row_to_csv()`

                    - `cleanup()`

        - NDO_TabDelimitedMatrix

            - nothing but an init - maybe just run the basic config through, make sure it comes out as we expect.

- // look at relation filtering/tie creation/rendering - Entity 10 (person 872) has two ties in test output for "basic" (to entities 8 and 9), should only have 1.  Article (21409) had two authors, so single article resulted in two ties to the subject, one from each author.
- test by comparing to output from the original tool, including derived statistics.

    - Jupyter notebooks:
    
        - overview of network data creation and analysis: [phd_work/methods/methods_paper_planning.ipynb#Network-Analysis](../methods/methods_paper_planning.ipynb#Network-Analysis)
        
            - old version of this information: [phd_work/methods/network_analysis/methods-network_analysis-create_network_data.ipynb](../methods/network_analysis/methods-network_analysis-create_network_data.ipynb)
        
        - R code for analysis comparing human and automated networks from original output: [phd_work/methods/network_analysis/statnet/R-statnet-grp_month-full_month.ipynb](../methods/network_analysis/statnet/R-statnet-grp_month-full_month.ipynb)
        - R notebook for new analysis: [phd_work/methods/network_analysis/statnet/R-statnet-grp_month-full_month-context_test.ipynb](../methods/network_analysis/statnet/R-statnet-grp_month-full_month-context_test.ipynb)
    
    - row counts are not the same, so need to look into what is going on.
    - spec for just automated: `./grp_month_from_context.json`
    
        - output file: `./grp_month_from_context.tsv`
        - attempt to just get automated coding from previous tool: `context_text_data-20200205-023708-just_automated.txt`
        - to start, probably should create a program to compare the two:
        
            - for each relation in one, then the other:
            
                - look up the article in question.
                - pull in data on it in context_text, and in context.
                - compare - how many relations in context, compared to Article_Data for that article in context_text for just the automated coder.
                - looking for two classes of problem: 1) where data was moved over incorrectly; and 2) where the filter criteria failed to pull in something in either system that should have been pulled in.
                - need to see why "just automated, just GRP, between 2009-12-01 and 2009-12-31" gives different results in the two systems.

## TODO - general

- Back to [Table of Contents](#Table-of-Contents)

**general TODO:**

- methods to find relations, similar to `filter_entities()` and `lookup_entities()` in `Entity` model class.  Include:

    - // Entity_Relation_Type by either slug or instance
    - // from = 
    - from_in
    - from_type_in
    - from_identifiers_in
    - from_entity_traits

        - AND dictionary
        - OR dictionary
        - fancy
    
    - // to = 
    - to_in
    - to_type_in
    - to_identifiers_in
    - to_entity_traits

        - AND dictionary
        - OR dictionary
        - fancy
    
    - // through =
    - through_in
    - through_identifiers_in
    - through_entity_traits

        - AND dictionary
        - OR dictionary
        - fancy
    
    - either FROM or TO (so undirected search - "I don't care which side")
    
        - IDs
        - identifiers
        - traits
    
    - any of FROM, TO, THROUGH
    
        - IDs
        - identifiers
        - traits
    
    - relation_traits
    
        - // AND dictionary
        - OR dictionary
        - fancy
    
    - _NOTE_ for trait matching types (and probably entity identifiers, also):

        - "AND dictionary" - to start, accept a dictionary of trait names and values that all must match (AND match).
        - "OR dictionary" - could add an "OR" trait match dictionary as well.
        - "fancy" - Eventually, could add ability to spec trait name (or unsaved model instance?), then specify test for value (equals, contains, etc.) and test value.  This would likely be a new little object.
        
    - Tags on Entity and/or Entity_Relation.

- make small dummy person and organization classes in context, so I can use them to test Abstract_Entity_Container without needing context_text.
- come up with better way to seed entities and relations for sourcenet - store spec in context_text base, then method to create or update all.

    - store the JSON from the context fixture in Context_Text_Base?  Or in a sourcenet class somewhere?
    - make a class in context/shared that contains variables and methods to loops over the items in context fixture JSON and update the database based on what is inside.  For each item in the fixture JSON, looks each up based on unique idnetifying information (name, slug, label, etc.). If it finds it, moves on.  If not, creates it.  
    - This will need to build up a basic object mapping based on foriegn keys before it creates anything, then create things in the right order (Entity_Types and related first, then Entity_Identifier_Types, then Relation_Types).  Order:
    
        - context.Trait_Type \
        - context.Entity_Type \
        - context.Entity_Type_Trait \
        - context.Entity_Identifier_Type
        - context.Entity_Relation_Type \
        - context.Entity_Relation_Type_Trait \
        - context.Term_Relation_Type \
        - context.Vocabulary \
        - context.Term \
        - context.Term_Relation \
        
    - for each type:
    
        - make a map of id to fields for each item of that type.
        - make a method for creating an instance of that type from fields.
        - to associate related, retrieve instance from in-memory map based on ID, then if db_id present, use it to look up, else look up in database based on name.  If not found, error, but could create.

    - To actually load, go in order of types outlined above, creating as you go.
    - When one of the items is added to the database, add a db_id field to their "fields".
    
- abstraction:

    - make an abstract parent for a type that has associated trait specs (parent to `Entity_Type` and `Entity_Relation_Type`).
    
        - share method `get_trait_spec()`.
    
    - make an abstract parent for trait containers that have associated types with associated trait specs (parent to `Entity` and `Entity_Relation`).
    
        - share method `

# DONE

- Back to [Table of Contents](#Table-of-Contents)

## DONE - TODO - Network data creation

- Back to [Table of Contents](#Table-of-Contents)

Network data creation TODO:

### DONE - TODO - Network data creation - framework

- Back to [Table of Contents](#Table-of-Contents)

Framework todos:

**_2020.01.06_**

- then, build basic framework where each output type accepts these two QuerySets, renders and returns desired output format.

    - refactoring from old NetworkOutput, NetworkDataOutput, and NDO objects:

        - // change so they refer to entities, not persons.
        - // pull out all of the Article_Data stuff.
        - // maintain the data structures used when actually rendering, just build from Entity_Relation, not Article_Data.  Might be able to get rid of some stuff, too.
        - // terms to search for and consider replacing "person" with "entity" (if they aren't in sections that will just be ripped out because they are no longer needed):
        
            - // person_dictionary
            - // person_dict
            - // article_data_query_set
            - // generate_master_person_list
            - // person_ids_list
            - // current_person_id 
            - // current_article_data
            - // article_data_counter
            - // master_person_list

        - todo:

            - // remove references to "include_render_details"
            - // remove "inclusion_params"
            
                - // remove references to `self.inclusion_params`
                - // remove references to `inclusion_params`
                - // remove references to `self.is_source_connected( current_source )`
            
            - // remove network_label
            - // update `generate_master_person_list()` to reference new variables (`generate_master_entity_list`).
            - // what to do about `get_person_label()`? - updated to `get_entity_label`, for now just uses Entity ID.  Could make it also include more IF Entity instances are cached.  We'll see if that is helpful.
            - // remove `get_person_type` and `get_person_type_id` (functions themselves are removed, need to mop up around the other classes: `grep -r -n "get_person_type" .`; `grep -r -n "get_person_type_id" .`
            - // rename `get_relations_for_person` to `get_relations_for_entity`
            - // rename `get_master_person_list` to `get_master_entity_list`
            - // rename `create_person_id_list` to `create_entity_id_list`
            - // rename `get_person_label` to `get_entity_label` (and made it a lot simpler).
            - // remove `PERSON_QUERY_TYPE_CHOICES_LIST` and related.
            - // remove `CODER_TYPE_FILTER_*`
            - // remove `PERSON_TYPE_*` variables (check in all files)
            - // search for "person", "people" in all NDO files in context.
            
                - // `ndo_simple_matrix.py`
                - // `ndo_csv_matrix.py`
                - // `ndo_tab_delimited_matrix.py`
            
            - // `Article_Subject`
            - // `append_person_row`...
            - // get rid of fancy date range code...
            - // remove `PARAM_*`?  Not for now - used in places, is a good signal for needing changes.
            
        - update person type stuff so it stores a list, rather than "author", "source", or "both".
        
            - // As you build data, keep track centrally of all relation types seen - `register_relation_type( relation_type_IN )`, called from `render()`, optionally also from `update_entity_relation_details()`.

**_2020.01.08_**

- build basic framework where each output type accepts these two QuerySets, renders and returns desired output format.

    - refactoring from old NetworkOutput, NetworkDataOutput, and NDO objects:

        - // `NetworkOutput.render_network_data()`: add outputting to file if file output path specified in request.
        - // add check to see if master list is larger after integrating items from network processing - if so, output error - for the entity filter to be effective, it must be a superset of the network set of relations.
        - // update person type stuff so it stores a list, rather than "author", "source", or "both".
        
            - // remove `update_person_type()`
            - // check for places that retrieve person type dictionary ( `self.person_type_dict` ) - removed all.
            - // create a variable, getter and setter for `relation_type_slug_to_instance_map` and `relation_type_slug_list`.
            - // removed `NDO.create_person_type_id_list` - figure out ramifications.  Might need to create a set of columns for each relation type, one for each of FROM, TO, and THROUGH, then populate appropriately from the new relation type map.    
            - // As you build data, keep track centrally of all relation types seen - call `register_relation_type( relation_type_IN )` to update map and list created above.  Updated from `render()`, optionally also from `update_entity_relation_details()`.
            - then, when outputting:
                
                - // for headers, grab this list of relation types and create FROM, TO, and THROUGH column headers for each.
                - // for tabular (`ndo_csv_matrix` and children):
                    
                    - // `NetworkDataOutput.create_relation_type_roles_for_entity()`: for data rows (attribute columns at right), walk the entity's relation type data structure in the same order as the relation type list, and output FROM, TO, and THROUGH numbers for each, 0 if not found.  Will result in many attribute columns.
                    - for data columns (attribute rows at bottom), pull in all relation types, then for each entity-->type-->role, walk all entities and output their value for that relation type in the row.  So, will result in many rows of attribute values.
                    
                        - // `NetworkDataOutput.create_relation_type_role_value_list()`: create method that accepts a relation type slug and a role, creates list of values for all entities in master entity list for that combination of slug and role.  If not present for a given entity, sets to 0.
                        - // `NetworkDataOutput.create_relation_type_value_dict()`: create a method that accepts a relation type slug, loops over all roles, calls `NetworkDataOutput.create_relation_type_role_value_list()` to build the list of values for each, then makes and returns dictionary mapping roles to value lists.
                        - // `NetworkDataOutput.create_all_relation_type_values_lists()`: create a method that loops over relation type slugs, then for each, calls `NetworkDataOutput.create_relation_type_value_dict()` to create dictionary that maps roles to values lists.  Creates a dictionary that maps relation type slugs to these dictionaries, then returns the new dictionary.
                        - // `NDO_CSVMatrix.append_entity_relation_type_rows()`: implement logic in the ndo_csv_matrix class that retrieves the values lists and uses them appropriately.
                    
                - // for `ndo_simple_matrix.py` (UCINet native format), need to implement `create_entity_relation_types_attribute_string()` - it assumed a single entity type value per person - need to re-do it so it pulls in all relation types, then outputs a list per person-->type-->role.  So, will result in many lists.

**2020.01.23**

- add ability in JSON to tell which entity traits we want to include in output traits, then when loading entities, look for those traits in each as it is loaded, store in a separate entity ID to trait name-value map.

    - // add place in output spec for:
    
        - traits (`output_entity_traits_list`?) - store the list of names/slugs of traits you want included if traits are output.  Possible filter criteria:
        
            - name
            - slug
            - entity_type_trait ID

        - identifiers (`output_entity_identifiers_list`) - list of identifiers you want to include - to effectively target, might need an object with more than just name.  Possible filter criteria:
        
            - name
            - id_type
            - source
            - identifier type ID
            
        - includes updating NetworkDataRequest and its tests to know of and allow for easy retrieval of these lists.

    - // add to processing a step in generating the master entity list where you loop over the list of traits if one present in the request and make a map of the values for those traits for each entity (map entity ID --> trait dict).
    
        - move the logic for processing entities to NetworkDataRequest, so it can be used to pass the entity dictionary, master entity list, and entry traits to NetworkDataOutput and children.
        
            - `process_entities`
            
                - `add_entities_to_dict`
                - `add_entity_to_dict`
                -  entity dictionary and trait map, and getters and setters.

                        self.m_entity_id_to_instance_map = {}
                        self.m_entity_id_to_traits_map = {}
                    
                - it checks if traits specified (call to `NetworkDataRequest.do_output_entity_traits_or_ids()`).
                - if so, calls `load_entities_traits_and_ids`.  Inside:
            
                    - `load_entity_traits`
                    - `load_entity_identifiers`
                
        - need to remove all that stuff from NetworkOutput, fix everything so it works again.
        - move all the test cases over to NetworkDataRequest.
    
    - // in rendering output, if traits output, update to render:
    
        - 1) entity ID
        - 2) entity relation type role information
        - 3) any requested traits

### DONE - TODO - Network data creation - testing

- Back to [Table of Contents](#Table-of-Contents)

Testing todos:

- unit testing:

    - // NetworkOutput

        - // `create_ndo_instance()`
        - // `get_NDO_instance()`
        - // `get_network_data_request()`
        - // `get_relation_query_set()`
        - // `set_NDO_instance()`
        - // `set_network_data_request()`
        - // `set_relation_query_set()`
        
    - NetworkDataOutput

        - // getters and setters

            - // `get_entity_dictionary()`/`set_entity_dictionary()`
            - // `get_entity_relation_type_summary_dict()`/`set_entity_relation_type_summary_dict()`
            - // `get_master_entity_list()`/`set_master_entity_list()`
            - // `get_network_data_request()`/`set_network_data_request()`
            - // `get_output_format()`/`set_output_format()`
            - // `get_output_structure()`/`set_output_structure()`
            - // `get_output_type()`/`set_output_type()`
            - // `get_query_set()`/`set_query_set()`
            - // `get_relation_map()`/`set_relation_map()`
            - // `get_relation_type_slug_list()`/`set_relation_type_slug_list()`
            - // `get_relation_type_slug_to_instance_map()`/`set_relation_type_slug_to_instance_map()`

        - // `set_query_set()`
        - // `set_entity_dictionary()`
        - // `initialize_from_request()`
        - // `render()`

            - // `get_query_set()`
            - // `get_entity_dictionary()`
            - // `add_directed_relation()`

                - // `get_relation_map()`

            - // `add_reciprocal_relation()`

                - // (duplicate) `add_directed_relation()`

            - // `register_relation_type()`

                - // `get_relation_type_slug_to_instance_map()`
                - // `get_relation_type_slug_list()`

            - // `update_entity_relations_details()`

                - // (duplicate) `register_relation_type()`
                - // (duplicate)`get_entity_relation_type_summary_dict()`

            - // `generate_master_entity_list()`

                - // `get_entity_dictionary()`
                - // (duplicate) `get_entity_relation_type_summary_dict()`
                - // (duplicate) `set_master_entity_list()`
                - // (duplicate) `get_master_entity_list()`

            - abstract `render_network_data()`

        - dependencies for child classes:

            - // `create_entity_id_list()`
            - // `create_all_relation_type_values_lists()`

                - // `get_relation_type_slug_list()`
                - // `create_relation_type_value_dict()`

                    - // `create_relation_type_role_value_list()`

                        - // (duplicate) `get_master_entity_list()`
                        - // `get_entity_relation_type_summary_dict()`
                        - // create goal data:
                        
                            - write program to, for basic, then entity_selection:
                            
                                - setup and render.
                                - retrieve relation QS.
                                - retrieve master entity ID list.
                                - for each relation type

                                    - for each role:
                                    
                                        - loop over ID list, and filter to count all relations where the current ID is in the selected type and role.

    - // NetworkDataRequest

        - // `process_entities()`

            - // `create_entity_dict()`

                - // `add_entities_to_dict()`

                    - // `add_entity_to_dict()`



## DONE - TODO - general

- Back to [Table of Contents](#Table-of-Contents)

Nothing yet...